### PR TITLE
Refactor and improve UT for PoP token verifier 

### PR DESCRIPTION
**Pull request by nehaPrakash98**
* The goal of this PR is to improve the test coverage. All of the happy path and error cases (where feasible) are covered by unit tests.
* In order to improve the readability of the tests, the test code that makes the test PoP tokens has been refactored. The existing code was hard to extend for new scenarios and hard to follow. The new code uses a builder pattern to create the tokens.

Previous test coverage was ~80%
Current test coverage is 91.5%

 
image

### CODE REVIEW

File: azure-coverage.out
Comment (julienstroheker) [inline azure-coverage.out:1]: We don't need this file as long with the html one right ?

Comment (nehaPrakash98) [reply azure-coverage.out:1]: Oh right, I'll remove them

File: auth/providers/azure/pop_tokenverifier_test.go [hunk 1]
```diff
@@ -243,6 +268,10 @@ func TestPopTokenVerifier_Verify(t *testing.T) {
 	_, err := verifier.ValidatePopToken(validToken)
 	assert.NoError(t, err)
 
+	validToken, _ = GeneratePoPToken(time.Now().Unix(), "testHostname", tsClaimsTypeString)
```
Comment (julienstroheker) [inline auth/providers/azure/pop_tokenverifier_test.go:271]: remind me what is that test compared to the previous one ?

Also no harm adding a description comment line on top of each test.

**Comment by AzureMarker** [comment]
Make sure to fix the DCO check:
https://github.com/kubeguard/guard/pull/371/checks?check_run_id=15500422959

File: auth/providers/azure/pop_tokenverifier_test.go (hunk 1 above)
Comment (nehaPrakash98) [reply auth/providers/azure/pop_tokenverifier_test.go]: Added comments for all tests

File: auth/providers/azure/pop_tokenverifier_test.go [hunk 2]
```diff
@@ -143,160 +144,330 @@ const (
...
-	var header, headerB64 string
-	nonce := uuid.New().String()
-	nonce = strings.Replace(nonce, "-", "", -1)
-	keyID := popKey.KeyID()
-	algo := popKey.Alg()
-	typ := "pop"
-	if kid == headerBadKeyID {
-		keyID = ""
+
+	payload := fmt.Sprintf(`{ "at" : "%s", "ts" : %d, "u": "%s", "cnf":{"jwk":%s}, "nonce":"%s"}`, at, b.ts, b.hostName, b.popKey.Jwk(), nonce)
```
Comment (julienstroheker) [inline auth/providers/azure/pop_tokenverifier_test.go:252]: Can this be a struct as well ?

File: auth/providers/azure/pop_tokenverifier_test.go [hunk 3]
```diff
@@ -24,6 +24,7 @@ import (
 	"encoding/base64"
```
Comment (julienstroheker) [inline auth/providers/azure/pop_tokenverifier_test.go:24]: Can the package be converted to be a `package azure_test` ?

File: auth/providers/azure/pop_tokenverifier_test.go [hunk 4]
```diff
@@ -143,160 +144,330 @@ const (
...
+	tsClaimsTypeUnknown = "tsClaimsTypeUnknown"
+	uClaimsWrongType    = "uClaimsWrongType"
+	signatureWrongType  = "signatureWrongType"
 )
 
-func GeneratePoPToken(ts int64, hostName, kid string) (string, error) {
-	if strings.Contains(kid, badTokenKey) {
-		return kid, nil
+// A struct that represents a PoP token
+type PoPToken struct {
```
Comment (julienstroheker) [inline auth/providers/azure/pop_tokenverifier_test.go:164]: Can we move all these helper struct / function into a dedicated file (`pop_tokenverifier_utils_test.go` ?) and just have the `TestPopTokenVerifier_Verify(t *testing.T)` func in this one ?

File: auth/providers/azure/pop_tokenverifier_test.go [hunk 5]
```diff
@@ -143,160 +144,330 @@ const (
...
-	finalToken := headerB64 + "." + payloadB64 + "." + signatureB64
+	finalToken := b.token.Header + "." + b.token.Payload + "." + b.token.Signature
 	return finalToken, nil
 }
 
 func TestPopTokenVerifier_Verify(t *testing.T) {
 	verifier := NewPoPVerifier("testHostname", 15*time.Minute)
 
-	validToken, _ := GeneratePoPToken(time.Now().Unix(), "testHostname", "")
+	validToken, _ := NewPoPTokenBuilder().SetTimestamp(time.Now().Unix()).SetHostName("testHostname").GetToken()
```
Comment (julienstroheker) [inline auth/providers/azure/pop_tokenverifier_test.go:344]: Can this be refactored to use a TDT (Table driven test) approach ?

```
func Test(t *testing.T) {
	testCases := []struct {
		desc	string
		
	}{
		{
			desc: "",
			
		},
	}
	for _, tC := range testCases {
		t.Run(tC.desc, func(t *testing.T) {
			
		})
	}
}
```

**Review by julienstroheker — CHANGES_REQUESTED** [review]
Overall LGTM, this is really good improvment.

I added comment about structure.

File: auth/providers/azure/pop_tokenverifier_test.go (hunk 2 above)
Comment (nehaPrakash98) [reply auth/providers/azure/pop_tokenverifier_test.go]: If we change this to a struct, then type of each of the fields will be fixed. But, some of the test cases are to identify errors when the wrong type of argument is passed. If we make it a struct then we won't be able to test those. For example: the right type of `ts` is int64, but in `TsClaimsTypeString` case, we pass `ts` as a string.

**Comment by julienstroheker** [comment]
@weinong are you able to trigger BUILD to validates the UT on this PR please ?

**Comment by julienstroheker** [comment]
@nehaPrakash98 please check the DCO

**Review by julienstroheker — APPROVED** [review]
LGTM

**Review by weinong — APPROVED** [review]
LGTM
